### ISSUE TITLE
Save editor contents before execution, for recovery on app crash.

### ISSUE THREAD

**Issue by roycetech** [Sequel-Ace/Sequel-Ace#773]
Today I was working on a query and when I tried to execute it, the app crashed and I lost the query I've written. It'd be good if those changes weren't lost and saved to a temporary file that is available when the app is re-opened.

**Comment by Jason-Morcos** [comment]
Yeah, state saving would be great. Unfortunately it's hard to address at this time without a bigger rewrite. We're definitely exploring it.

As to your query, you may find it saved in the query console. We try really hard to save everything to the console when things go wrong!

Duplicates #729

**Comment by Jason-Morcos** [comment]
If the query isn't in the console, then it means it's not saving on crash to the console (which it probably should). So if it's not in your console, let us know and we'll re-open the issue! Thanks!

**Comment by roycetech** [comment]
No, it's not in the console, unfortunately, at least not the query that I just run that (likely) triggered the crash. I did see some queries related to the schema of DB I was working on, like SHOW DB, SHOW FULL TABLES etc. I checked the history as well, the latest query before it crashed wasn't there either.

**Comment by Jason-Morcos** [comment]
@Sequel-Ace/all Sounds like we may want to try to guard as much as possible to try to save any query to the console before crashing if an exception is thrown. Obviously will be hard if we crash in preparing the query, but worth a shot.
Also maybe writing the query editor contents to the console on crash too.

Perhaps additionally if we catch an exception we could write that we crashed to defaults before dying and if that wasCrashed default was set we could pop up a message to the user to check the console for queries they may have lost.

**Comment by jamesstout** [comment]
I sort of made a start: https://github.com/Sequel-Ace/Sequel-Ace/tree/store-data-before-crash

**Comment by jamesstout** [comment]
Alternatively: https://github.com/Sequel-Ace/Sequel-Ace/tree/catch-crashes

**Comment by jamesstout** [comment]
Actually, started thinking that maybe we should save the spfs session file, including query text, every few mins?

**Comment by Jason-Morcos** [comment]
I like both approaches! Maybe a combo of the 2? Could we swizzle the crash to do a save and then allow the exception to bubble up and explode? Maybe tell the user we had an exception and we can try to popup an alert asking if they want to continue or allow the app to crash?

I kinda like the auto-save idea too, James. Perhaps, lean more towards the first approach you wrote (the store data before crash) and start with that and go from there adding other things later only if needed?
